harness/tests/launch/test_autodeepspeed.py

```diff
@@ -0,0 +1,229 @@
+        "-m",
+        "determined.exec.pid_server",
+        "--on-fail",
+        "SIGTERM",
+        "--on-exit",
+        "SIGTERM",
+        "--grace-period",
+        "5",
+        f"/tmp/pid_server-{cluster_info.allocation_id}",
+        str(len(cluster_info.slot_ids)),
+        "--",
+        "deepspeed",
+        "-H",
+        autodeepspeed.hostfile_path,
+        "--master_addr",
+        "localhost",
+        "--no_python",
+        "--no_local_rank",
+        "--",
+        "python3",
+        "-m",
+        "determined.exec.pid_client",
+        f"/tmp/pid_server-{cluster_info.allocation_id}",
+        "--",
+        "python3",
+        "-m",
+        "determined.exec.worker_process_wrapper",
+        "RANK",
+        "--",
+        "python3",
+        "-m",
+        "determined.exec.harness",
+        "--train-entrypoint",
+        train_entrypoint,
+    ]
+
+    with set_env_vars():
+        autodeepspeed.main(train_entrypoint)
+
+    mock_cluster_info.assert_called_once()
+    assert os.environ["DET_CHIEF_IP"] == cluster_info.container_addrs[0]
+    assert os.environ["USE_DEEPSPEED"] == "1"
+
+    mock_subprocess.assert_called_once_with(launch_cmd)
+
+
+@patch("subprocess.Popen")
+@patch("determined.get_cluster_info")
+@patch("determined.common.api.post")
+def test_launch_worker(
+    mock_api: MagicMock, mock_cluster_info: MagicMock, mock_subprocess: MagicMock
+) -> None:
+    cluster_info = get_mock_cluster_info(["0.0.0.0", "0.0.0.1"], 1)
+    mock_cluster_info.return_value = cluster_info
+    with set_env_vars():
+        autodeepspeed.main("model_def:TrialClass")
+
+    mock_cluster_info.assert_called_once()
+    assert os.environ["DET_CHIEF_IP"] == cluster_info.container_addrs[0]
+
+    mock_api.assert_called_once()
+    expected_cmd = [
+        "python3",
+        "-m",
+        "determined.exec.pid_server",
+        "--on-fail",
+        "SIGTERM",
+        "--on-exit",
+        "SIGTERM",
+        "--grace-period",
+        "3",
+        f"/tmp/pid_server-{cluster_info.allocation_id}",
+        str(len(cluster_info.slot_ids)),
+        "--",
+        "/usr/sbin/sshd",
+        "-p",
+        str(constants.DTRAIN_SSH_PORT),
+        "-f",
+        "/run/determined/ssh/sshd_config",
+        "-D",
+    ]
+    mock_subprocess.assert_called_once_with(expected_cmd)
+
+
+def get_mock_cluster_info(container_addrs: List[str], container_rank: int) -> ClusterInfo:
+    rendezvous_info_mock = RendezvousInfo(
+        container_addrs=container_addrs, container_rank=container_rank
+    )
+    cluster_info_mock = ClusterInfo(
+        master_url="localhost",
+        cluster_id="clusterId",
+        agent_id="agentId",
+        slot_ids=[0, 1, 2, 3],
+        task_id="taskId",
+        allocation_id="allocationId",
+        session_token="sessionToken",
+        task_type="TRIAL",
+        rendezvous_info=rendezvous_info_mock,
+    )
+    return cluster_info_mock
+
+
+@contextlib.contextmanager
+def set_env_vars() -> Iterator[None]:
```

Comment:
rename to `set_container_id_env_var` or something, so its name matches what it does